core/storage/btree.rs

```diff
@@ -1457,28 +1461,33 @@ impl BTreeCursor {
             let mut size = 0;
             loop {
                 // TODO: check corruption icellast
-                next = u16::from_be_bytes(buf[pc..pc + 2].try_into().unwrap()) as usize;
-                size = u16::from_be_bytes(buf[pc + 2..pc + 4].try_into().unwrap()) as usize;
+                next = u16::from_be_bytes(buf[pc..pc + 2].try_into().unwrap()) as usize; // first 2 bytes in freeblock = next freeblock pointer
+                size = u16::from_be_bytes(buf[pc + 2..pc + 4].try_into().unwrap()) as usize; // next 2 bytes in freeblock = size of current freeblock
                 nfree += size;
+                // Freeblocks are in order from left to right on the page,
+                // so next pointer should > current pointer + its size, or 0 if no next block exists.
                 if next <= pc + size + 3 {
                     break;
                 }
                 pc = next;
             }
 
-            if next > 0 {
-                todo!("corrupted page ascending order");
-            }
+            // Next should always be 0 (NULL) at this point since we have reached the end of the freeblocks linked list
+            assert!(
+                next == 0,
+                "corrupted page: freeblocks list not in ascending order"
+            );
 
-            if pc + size > usable_space {
-                todo!("corrupted page last freeblock extends last page end");
-            }
+            assert!(
+                pc + size <= usable_space,
+                "corrupted page: last freeblock extends last page end"
+            );
         }
 
         // if( nFree>usableSize || nFree<iCellFirst ){
         //   return SQLITE_CORRUPT_PAGE(pPage);
         // }
-        // don't count header and cell pointers?
+        // FIXME:don't count header and cell pointers?
```

Comment:
i think this is probably indeed a current bug. page layout is like

```
HEADER -- CELL POINTERS -- UNALLOC -- CELL CONTENT AREA -- RESERVED
```

so I think it makes sense we shouldnt count header and cell pointer array towards free space

Reply:
Never mind, the existing code does:

```
nfree -= first_cell as usize;
```

i'll rename `first_cell` or refactor this somehow to make it clearer